sansa-ml-spark/src/main/scala/net/sansa_stack/ml/spark/clustering/algorithms/LocalitySentiveHashing.scala

```diff
@@ -0,0 +1,144 @@
+    val hashingTf = new HashingTF().setInputCol("filtered_words").
+      setOutputCol("raw_Features").setNumFeatures(Math.round(0.90 * vocab_size).toInt)
+    val isNonZeroVector = udf({ v: Vector => v.numNonzeros > 0 }, DataTypes.BooleanType)
+    val featuredData_hashedDf = hashingTf.transform(cleanData_Df).filter(isNonZeroVector(col("raw_Features")))
+    val idf = new IDF().setInputCol("raw_Features").setOutputCol("features")
+    val idfModel = idf.fit(featuredData_hashedDf)
+    val rescaledHashedData = idfModel.transform(featuredData_hashedDf).
+      filter(isNonZeroVector(col("features")))
+
+    return rescaledHashedData
+  }
+
+  def calculateVocabsize(cleandfrdd: RDD[Row]): Int = {
+    val vocab = cleandfrdd.map(_.mkString).reduce(_ + ", " + _).split(", ").toSet
+    return (vocab.size)
+
+  }
+
+  def minHashLSH(featuredData_Df: DataFrame): (MinHashLSHModel, DataFrame) = {
+    val mh = new MinHashLSH().setNumHashTables(3).setInputCol("features").setOutputCol("HashedValues")
+    val model = mh.fit(featuredData_Df)
+    val transformedData_Df = model.transform(featuredData_Df)
+    return (model, transformedData_Df)
+  }
+
+  def approxSimilarityJoin(model: MinHashLSHModel, transformedData_Df: DataFrame): Dataset[_] = {
+    val threshold = 0.40
+    return model.approxSimilarityJoin(transformedData_Df, transformedData_Df, threshold)
+  }
+
+  def clusterFormation(dataset: Dataset[_], featuredData_Df: DataFrame) = {
+    val featuredData = featuredData_Df.drop("attributes", "words", "filtered_words")
+
+    val refined_entities_dataset = dataset
+      .filter("datasetA.entities != datasetB.entities")
+      .select(col("datasetA.entities").alias("src"), col("datasetB.entities").alias("dst"))
+
+    import spark.implicits._
+    val c_1 = refined_entities_dataset.select("src")
+    val c_2 = refined_entities_dataset.select("dst")
+    val vertexDF = c_1.union(c_2).distinct().toDF("id")
+
+    val g = GraphFrame(vertexDF, refined_entities_dataset)
+    g.persist()
+    spark.sparkContext.setCheckpointDir(dir_path)
+    
+    // Connected Components are the generated clusters.
+    val connected_components = g.connectedComponents.run()
+    val connected_components_ = connected_components.withColumnRenamed("component", "prediction").
+      withColumnRenamed("id", "entities")
+    clusterQuality(connected_components_, featuredData)
+  }
+  
+  /*
+   * Calculating Silhouette score, which will tell how good clusters are.
+   * Silhouette values ranges from [-1,1]. 
+   * Values closer to 1 indicates better clusters
+   */
+  
+  def clusterQuality(connectedComponents: DataFrame , featuredData: DataFrame) = {
+    var silhouetteInput = connectedComponents.join(featuredData, "entities")
+    val evaluator = new ClusteringEvaluator().setPredictionCol("prediction").
+      setFeaturesCol("features").setMetricName("silhouette")
+    val silhouette = evaluator.evaluate(silhouetteInput)
+    println(s"Silhouette for LSH = $silhouette")    
```

Comment:
We should exclude `println` statements here. Considering adding unit-tests which may include the clsuter-quality-chech as well (on sample data)